### DISCUSSION TITLE
Webhook https

### DISCUSSION THREAD

**Discussion by laksa19** [aldinokemal/go-whatsapp-web-multidevice#158] [General]
Saya coba tes menggunakan webhook https, namun muncul error "connect: connection refused"

**Comment by aldinokemal** [comment]
`connect: connection refused` itu error di level TCP, jadi bukan masalah HTTPS/sertifikat. Artinya proses gowa berhasil resolve host tujuan tapi tidak ada yang listen di port itu (atau firewall menolak koneksi), dan koneksi ditolak sebelum TLS handshake dimulai.

Yang perlu dicek:

1. Tes dari mesin/container yang sama dengan gowa, bukan dari laptop:
   ```
   curl -v -X POST https://yourcallback.com/callback
   ```
   Kalau curl juga `connection refused`, masalahnya di jaringan/endpoint, bukan di aplikasi.
2. Kalau gowa jalan di Docker dan webhook diarahkan ke `localhost` / `127.0.0.1`, itu menunjuk ke dalam container gowa sendiri. Pakai `host.docker.internal` (Docker Desktop) atau nama service di compose.
3. URL `https://` tanpa port berarti port 443. Kalau endpoint Anda listen di port lain (misal 8443), portnya harus ditulis eksplisit.

Kalau nanti errornya berubah jadi `x509: certificate signed by unknown authority` (self-signed cert), itu baru urusan TLS dan ada opsinya di `src/infrastructure/whatsapp/webhook.go:35-39`:
```
--webhook-insecure-skip-verify=true
```
Bisa juga lewat env `WHATSAPP_WEBHOOK_INSECURE_SKIP_VERIFY=true`.

FYI, setiap event webhook yang gagal di-retry 5x dengan backoff (`src/infrastructure/whatsapp/webhook.go:64-85`), jadi satu event gagal akan muncul 5 baris warning di log. Itu normal, bukan 5 error terpisah.
